idx/memory/meta_tags.go

```diff
@@ -0,0 +1,236 @@
+	res := make([]string, len(m.metaTags))
+
+	for i, tag := range m.metaTags {
+		tag.stringIntoBuilder(builder)
+		res[i] = builder.String()
+		builder.Reset()
+	}
+
+	return res
+}
+
+func (m *metaTagRecord) queryStrings(builder *strings.Builder) []string {
+	res := make([]string, len(m.queries))
+
+	for i, query := range m.queries {
+		query.stringIntoBuilder(builder)
+		res[i] = builder.String()
+		builder.Reset()
+	}
+
+	return res
+}
+
+// hashQueries generates a hash of all the queries in the record
+func (m *metaTagRecord) hashQueries() uint32 {
+	builder := strings.Builder{}
+	for _, query := range m.queries {
+		query.stringIntoBuilder(&builder)
+
+		// trailing ";" doesn't matter, this is only hash input
+		builder.WriteString(";")
+	}
+
+	h := queryHash()
+	h.Write([]byte(builder.String()))
+	return h.Sum32()
+}
+
+// sortQueries sorts all the queries first by key, then by value, then by
+// operator. The order doesn't matter, it only needs to be consistent
+func (m *metaTagRecord) sortQueries() {
+	sort.Slice(m.queries, func(i, j int) bool {
+		if m.queries[i].key == m.queries[j].key {
+			if m.queries[i].value == m.queries[j].value {
+				return m.queries[i].operator < m.queries[j].operator
+			}
+			return m.queries[i].value < m.queries[j].value
+		}
+		return m.queries[i].key < m.queries[j].key
+	})
+}
+
+// matchesQueries compares another tag record's queries to this
+// one's queries. Returns true if they are equal, otherwise false.
+// It is assumed that all the queries are already sorted
+func (m *metaTagRecord) matchesQueries(other metaTagRecord) bool {
+	if len(m.queries) != len(other.queries) {
+		return false
+	}
+
+	for i, query := range m.queries {
+		if query.key != other.queries[i].key {
+			return false
+		}
+
+		if query.operator != other.queries[i].operator {
+			return false
+		}
+
+		if query.value != other.queries[i].value {
+			return false
+		}
+	}
+
+	return true
+}
+
+// hasMetaTags returns true if the meta tag record has one or more
+// meta tags, otherwise it returns false
+func (m *metaTagRecord) hasMetaTags() bool {
+	return len(m.metaTags) > 0
+}
+
+// upsert inserts or updates a meta tag record according to the given specifications
+// it uses the set of tag query expressions as the identity of the record, if a record with the
+// same identity is already present then its meta tags get updated to the specified ones.
+// If the new record contains no meta tags, then the update is equivalent to a delete.
+// Those are the return values:
+// 1) The id at which the new record got inserted
+// 2) Pointer to the inserted metaTagRecord
+// 3) The id of the record that has been replaced if an update was performed
+// 4) Pointer to the metaTagRecord that has been replaced if an update was performed, otherwise nil
+// 5) Error if an error occurred, otherwise it's nil
+func (m metaTagRecords) upsert(metaTags []string, tagQueryExpressions []string) (uint32, *metaTagRecord, uint32, *metaTagRecord, error) {
```

Comment:
I have moved things around a bit: https://github.com/grafana/metrictank/pull/1301/commits/dd5f17ae0b16edc91af38fd9739958e063247e1e

I'm not sure where to put `metaTagIndex` with its methods though, as it's kind of independent from the other stuff in this file. Do you think it would be better to put `metaTagIndex` and its methods into their own file?

Reply:
can stay where it is, but why are the methods for metaTagRecords (actually just one) so far away from the type declaration?
after a type/structure, we should put its constructor ( if any), then its methods.
if this is not clear in the rule, help me clarify it.

Reply:
If I move the `metaTagRecords` method (`upsert()`) up so it is right below the type definition, then the the `metaTagIndex` / `metaTagValue` type definitions would be below the `upsert()` method. 
Now that I re-read the point 12 i'm realizing there's nothing that says that all type definitions should be at the top of the file, before the method implementations, I guess I just assumed that out of habit.

Reply:
How's this: https://github.com/grafana/metrictank/pull/1301/commits/4df746a28cab60f8e61b4f974c8f1ea77da91854

To me it feels weird to put type definitions into the middle of the file, but I'll get used to that.

Reply:
> To me it feels weird to put type definitions into the middle of the file, but I'll get used to that.

it's a golang best practice. check any stdlib file and it's like this, e.g. https://golang.org/src/net/http/header.go